File: DSA/ArrayHashing/TwoSum/TwoSum.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

std::vector<int> twoSum(std::vector<int> &nums, int target)
{
    std::unordered_map<int, int> map;
    std::vector<int> result;

    for (int i = 0; i < nums.size(); i++)
    {
        int diff = target - nums[i];

        if (map.find(diff) != map.end())
        {
            result.push_back(map[diff]);
            result.push_back(i);
            return result;
        }

        else
        {
            map.insert({nums[i], i});
        }
    }

    return {-1, -1};
}

std::ostream &operator<<(std::ostream &os, const std::vector<int> &input)
{
    os << "[ ";
    for (auto const &i : input)
    {
        os << i << " ";
    }

    os << "]";

    return os;
}

int main(int argc, char *argv[])
{
    std::vector<int> arr1 = {10, 7, 2, 15, 1};
    std::vector<int> arr2 = {2, 5, 5, 1};

    std::cout << twoSum(arr1, 17) << std::endl;
    std::cout << twoSum(arr1, 9) << std::endl;
    std::cout << twoSum(arr1, 25) << std::endl;
    std::cout << twoSum(arr1, 16) << std::endl;
    std::cout << twoSum(arr1, 100) << std::endl;
    std::cout << twoSum(arr2, 10) << std::endl;
}

File: DSA/ArrayHashing/TwoSum/twoSum.js

const twoSum = (nums, target) => {
    const map = new Map();

    for (let i = 0; i < nums.length; i++) {
        const num = nums[i];
        const diff = target - num;

        if (map.has(diff)) {
            return [map.get(diff), i]
        }

        else {
            map.set(num, i);
        }
    }

    return [-1, -1];
}




const arr1 = [10, 7, 2, 15, 1]
const arr2 = [2, 5, 5, 1]


console.log(twoSum(arr1, 17)) // [0, 1]
console.log(twoSum(arr1, 9))  // [1, 2]
console.log(twoSum(arr1, 25))  // [0, 3]
console.log(twoSum(arr1, 16))  // [3, 4]
console.log(twoSum(arr1, 100))  // [-1, -1]
console.log(twoSum(arr2, 10))  // [ 1, 2]